docs/content/preview/architecture/docdb-replication/raft.md

```diff
@@ -0,0 +1,74 @@
+
+1. The leader first takes a lock on the key using a local in-memory lock manager and reads data if necessary (for read-modify-write or conditional update operations). Note that this locking mechanism does not exist on followers.
+1. The batch of changes to be written to [DocDB](../../docdb/) is prepared. This write batch is very close to the final set of RocksDB key-value pairs to be written, only lacking the final [hybrid timestamp](../../transactions/transactions-overview#hybrid-logical-clocks) at the end of each key.
+1. The leader appends the batch to its Raft log and chooses a [hybrid timestamp](../../transactions/transactions-overview#hybrid-logical-clocks) for the write operation.
+1. The data is replicated via Raft to its peers.
+1. After receiving a `replication successful` callback from the Raft subsystem, the leader applies the write batch to its local DocDB.
+1. Responds with success to the user.
+1. The next update from the leader notifies followers that the entry has been committed, and the followers apply the write batch to their RocksDB instances.
+
+## Leader election
+
+When the system starts up, each node is assigned an initial election term, which is typically a non-negative integer. When a  follower does not receive a heartbeat within the election timeout period it becomes a candidate to start a new election term and attempt to become the new Leader. Each node votes for a candidate in a particular election term. The candidate that receives votes from a majority of the nodes is elected as the new leader. Whenever a new leader is elected, the election term is incremented. This ensures that each term has a unique identifier, and it helps the nodes in the system differentiate between the various terms.
+
+If a node receives a request from a leader with a stale term (i.e., a term that is lower than the node's current term), the node will reject the request and inform the leader of the newer term. This helps prevent outdated leaders from making changes to the replicated log. RAFT ensures that the replicated log remains consistent across the nodes by requiring that all log entries be associated with a specific term. This means that a node will only accept log entries that are from the current term or a previous term
+
+The Leader continues to send heartbeat messages to maintain its leadership status. If a Follower does not receive a heartbeat within the election timeout period, a new election is triggered.
+
+## Consistency guarantees
+
+RAFT ensures that the replicated log remains consistent across all the nodes in the cluster. It achieves this by enforcing a rule that the leader can only append new entries to the log, and it must ensure that those entries are consistent with the log of the majority of the nodes. If a follower's log is not consistent with the leader's log, the leader will override the follower's log with its own.
```

Comment:
cluster->universe.  The read replicas are also part of the Raft groups
only append->append only
the leader can remove entries at the end of its log that are not majority replicated
new entries don't have consistency issues with the nodes because they won't be shared by those nodes


actually, I'm not sure what you mean by "append new entries to the log".  You mean to the local log of the leader or to the virtual log defined by the majority replicated entries in all the nodes?

this paragraph needs work

Reply:
actually this should be Raft group instead of cluster as well
I tend to use the word participant for the members of a raft group rather than nodes